src/freertos_drivers/common/MCP2515Can.hxx

```diff
@@ -584,7 +594,141 @@ private:
      */
     MCP2515Can();
 
+    /** Allow access to MCP2515Can from MCP2515GPO */
+    friend class MCP2515GPO;
+
+    /** Allow access to MCP2515Can from MCP2515GPI */
+    friend class MCP2515GPI;
+
     DISALLOW_COPY_AND_ASSIGN(MCP2515Can);
 };
 
+/** General Purpose Output (GPO) instance on the MCP2515.
+ */
+class MCP2515GPO : public Gpio
+{
+public:
+    /** Constructor.
+     * @param instance parrent MCP2515Can instance that "owns" the interface.
+     * @param bit bit index (0 through 1) of the output.
+     */
+    MCP2515GPO(MCP2515Can *instance, uint8_t bit)
+        : Gpio()
+        , instance_(instance)
+        , bit_(bit)
+    {
+        HASSERT(bit < 2);
+    }
+
+    /** Writes a GPO pin (set or clear to a specific state).
+     * @param new_state the desired output state.  See @ref Value.
+     */
+    void write(Value new_state) const override
+    {
+        new_state ? set() : clr();
+    }
+
+    /** Retrieves the current @ref Value of a GPO output sate (requested).
+     * @return @ref SET if currently high, @ref CLR if currently low
+     */
+    Value read() const override
+    {
+        return instance_->gpoData & (0x1 << bit_) ? Gpio::SET : Gpio::CLR;
+    }
+
+    /** Sets the GPO pin to high.
+     */
+    void set() const override
+    {
+        portENTER_CRITICAL();
+        instance_->gpoData |= 0x1 << bit_;
+        portEXIT_CRITICAL();
+        instance_->sem.post();
+    }
+
+    /** Clears the GPO pin to low.
+     */
+    void clr() const override
+    {
+        portENTER_CRITICAL();
+        instance_->gpoData &= ~(0x1 << bit_);
+        portEXIT_CRITICAL();
+        instance_->sem.post();
+    }
+
+    /** Sets the GPO direction (does nothing).
+     * @param dir @ref INPUT or @ref OUTPUT
+     */
+    void set_direction(Gpio::Direction dir) const override
+    {
+    }
+
+    /** Gets the GPO direction.
+     * @return always returns @ref OUTPUT
+     */
+    Direction direction() const override
+    {
+        return Gpio::Direction::OUTPUT;
+    }
+
+private:
+    /** reference to chip instance */
+    MCP2515Can *instance_;
+
+    /** bit number representative of the bit */
+    uint8_t bit_;
+
+    DISALLOW_COPY_AND_ASSIGN(MCP2515GPO);
+};
+
+/** General Purpose Input (GPI) instance on the MCP2515.
+ */
+class MCP2515GPI : public Gpio
+{
+public:
+    /** Constructor.
+     * @param instance parrent MCP2515Can instance that "owns" the interface.
+     * @param bit bit index (0 through 2) of the input.
+     */
+    MCP2515GPI(MCP2515Can *instance, uint8_t bit)
+        : Gpio()
+        , instance_(instance)
+        , bit_(bit)
+    {
+        HASSERT(bit < 3);
+    }
+
+    /** Retrieves the current @ref Value of a GPI input pin.
+     * @return @ref SET if currently high, @ref CLR if currently low
+     */
+    Value read() const override
+    {
+        return instance_->gpiData & (0x1 << bit_) ? Gpio::SET : Gpio::CLR;
```

Comment:
how about we post the semaphore once for each read? Or if the last read was older than 20 msec, we post?
This would allow to poll exactly at the frequency that the userspace code polls, they will just be always one tick late.
Including never polling if noone reads.
Also if we post the semaphore, and the priorities are appropriately set up, then our current thread will be interrupted and by the time we come back to the CPU the new value will be in gpiData.

Reply:
I added the post, but I did not add the timeout (for now).